File: adventofcode-2016/src/main/java/nl/mvdr/adventofcode/adventofcode2016/day11/Part2.java
package nl.mvdr.adventofcode.adventofcode2016.day11;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.solver.Solver;

public class Part2 implements Solver {

    private static final Logger LOGGER = LoggerFactory.getLogger(Part2.class);
    
    @Override
    public String solve(String inputfile) {
        return new RadioisotopeThermoelectricGenerators(true).solve(inputfile);
    }
    
    public static void main(String[] args) {
        Part2 instance = new Part2();

        String result = instance.solve("input-day11-2016.txt");

        LOGGER.info(result);
    }
}

File: adventofcode-2025/src/main/kotlin/nl/mvdr/adventofcode/adventofcode2025/day11/Part2.kt
package nl.mvdr.adventofcode.adventofcode2025.day11

import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.solver.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart2(lines: Sequence<String>): Long {
    val algorithm = createAlgorithm(lines)

    val svrToDacToFftToOut = when (val dacToFft = algorithm.countAllPaths("dac", "fft")) {
        0L -> 0L
        else -> {
            val svrToDac = algorithm.countAllPaths("svr", "dac")
            val fftToOut = algorithm.countAllPaths("fft", "out")
            svrToDac * dacToFft * fftToOut
        }
    }

    val svrToFftToDacToOut = when(val fftToDac = algorithm.countAllPaths("fft", "dac")) {
        0L -> 0L
        else -> {
            val svrToFft = algorithm.countAllPaths("svr", "fft")
            val dacToOut = algorithm.countAllPaths("dac", "out")
            svrToFft * fftToDac * dacToOut
        }
    }

    return svrToDacToFftToOut + svrToFftToDacToOut
}

fun main() {
    val result = FunctionSolver(::solvePart2).solve("input-day11.txt")
    logger.info { result }
}
